File: built-in plugins dsp/mb-surround-encoder.cpp
/**
 * MB Surround Encoder
 * Category : effect
 * Type     : stereo
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Stereo to surround upmix encoder for immersive formats
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SURROUND_ENCODER_H
#define MB_SURROUND_ENCODER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSurroundEncoder : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-surround-encoder";
    static constexpr const char* PLUGIN_NAME    = "MB Surround Encoder";
    static constexpr const char* PLUGIN_TYPE    = "stereo";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float frontWidth = 0.8f;  // range [0, 1]
    float rearLevel = -6f;  // range [-24, 0]
    float lfeLevel = -10f;  // range [-24, 0]
    float centerLevel = -3f;  // range [-24, 0]
    };

    MbSurroundEncoder() = default;
    ~MbSurroundEncoder() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.frontWidth = std::clamp(params.frontWidth, 0f, 1f);
        params.rearLevel = std::clamp(params.rearLevel, -24f, 0f);
        params.lfeLevel = std::clamp(params.lfeLevel, -24f, 0f);
        params.centerLevel = std::clamp(params.centerLevel, -24f, 0f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Surround Encoder
        return input;
    }
};

#endif // MB_SURROUND_ENCODER_H

File: built-in plugins dsp/mb-surround-encoder.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbSurroundEncoderPlugin: PluginDefinition = {
  id: "mb-surround-encoder",
  slug: "mb-surround-encoder",
  name: "MB Surround Encoder",
  category: "effect",
  type: "stereo" as any,
  version: "1.0.0",
  description: "Stereo to surround upmix encoder for immersive formats",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "frontWidth",
      name: "Front Width",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "rearLevel",
      name: "Rear Level",
      type: "float",
      defaultValue: -6,
      minValue: -24,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "lfeLevel",
      name: "LFE Level",
      type: "float",
      defaultValue: -10,
      minValue: -24,
      maxValue: 0,
      automatable: true,
    },
    {
      id: "centerLevel",
      name: "Center Level",
      type: "float",
      defaultValue: -3,
      minValue: -24,
      maxValue: 0,
      automatable: true,
    },
  ],
  defaultPreset: {
    frontWidth: 0.8,
    rearLevel: -6,
    lfeLevel: -10,
    centerLevel: -3,
  },
};

export default MbSurroundEncoderPlugin;
